### ISSUE TITLE
jQuery Fundamentals

### ISSUE THREAD

**Issue by rrgayhart** [turingschool/intermission-assignments#69]
Respond in the comments of this issue with:
1. What is something you learned that was particularly surprising/interesting?
2. What was something you already knew?
3. Was there anything you feel you still don't understand?

Feel free to ask any questions about jQuery fundamentals here as well.

**Comment by NickyBobby** [comment]
1. Something I learned is that AJAX calls do not deal directly with API's. There is something similar that can be done with certain API's called JSONP. I look forward to exploring these concepts soon.
2. Something that I already knew was how to target certain tags and manipulate them in certain ways. I wouldn't say I know how to do this well but I have a pretty good idea of how it works. I also look forward to exploring these concepts soon. 
3. I guess what I'm still confused by is that since JQuery can only run in the console, where do we write this code? And does it get loaded into the console every time we refresh the page?

**Comment by julyytran** [comment]
1. It's safer to assign function expressions to variables versus using a function declaration. I had inconsistency in this with my personal project and was frequently confused by why I couldn't access certain functions/variables when I moved them. Maybe something about hoisting? 
2. I already knew how to manipulate DOM elements with jQuery, particularly using things like .hide(), .show(), and .toggleClass()
3. What does high-quality, refactored code look like in a project involving a lot of JS? What exactly is hoisting and how can I use it to my advantage? How exactly do you deal with a CORS issue when you don't control the server and they don't return JSONP?

**Comment by cheljoh** [comment]
What is something you learned that was particularly surprising/interesting?
- I Didn’t fully understand ‘this’ until reading this- refers to object in context in which the function was called
- That everything is an object except for- strings, booleans, numbers, undefined, and null
- Undefined, null, NaN, 0, and ‘’ are falsy
- Ternary operator: var propertyName = ( dim === 'width' ) ? 'clientWidth' : 'clientHeight’;
- $() function always returns something truthy
- filtering:
  var listItems = $( 'li' );
  var special = listItems.filter( '.special' );
- Can change attributes of elements- $( 'a' ).attr( 'title', 'Click me!' )

What was something you already knew?
- That any variable not declared with “var” is global
- You can access objects like this: person.firstName, or this: person[ 'lastName’ ]
- Objects can have methods, called like this person.greet(person.firstName)
- toggling:
  $( 'li' ).eq( 1 ).toggleClass( 'hidden' );

Was there anything you feel you still don't understand?
- Getters and setters
- Event bubbling/delegation

**Comment by stevekinney** [comment]
@NickyBobby jQuery doesn't have to run in the console and 99.9999999% of the time, it doesn't. At the simplest level, it runs between `<script>…</script>` in the browser. Rails includes it by default in the Asset Pipeline.

**Comment by stevekinney** [comment]
@julyytran Assigning a function to a variable and declaring a function are like 90% the same. The issue—of course—is the 10% in which they are different. In Ruby, when you load a file, it doesn't just start executing it from the top, it first scans the entire file and everything it requires. It looks at all the classes and the modules and the globals and registers all of them. Then and only then does Ruby begin to execute your file.

When we declare a function using a name, like so:

``` js
function someAmazingFunction() { return true; }
```

…it is hoisted to the top just like any Ruby class. But, also like Ruby, you can't use a variable before you assign it. So, this code will fail spectacularly:

``` js
someAmazingFunction();
var  someAmazingFunction = function () { return true; }
```

**Comment by natevenn** [comment]
Something that I learned was that you can use namespaces with .on() which allow you to target a specific event - example being $('li').on('click.logging', function(){ some code })

Something that I already knew was that everything in JS is truthy execpt null, zero, undefined, NaN, and empty strings

I don't entirely understand how deferred works and why you would use it. Looks like its just another way to manage asynchronous code and reduce deeply nested functions but its still confusing to me when I would know to use it and how.

**Comment by adriennedomingus** [comment]
Odd that an empty string is not truthy, but empty arrays, etc., are
Some of the math on non-number values aren't the behavior I'd expect
Setters work on all matching elements, getters return the first matching element
The idea of using .trigger() to trigger jQuery bindings, but not native behavior
Serializing form data in an AJAX request

I was already pretty comfortable traversing the DOM. The idea of checking the length of an object, not just that it exists before running a section of code was interesting though. Pretty sure that caused me some problems last project because the code always ran even when the element didn't exist and I wasn't checking for length :)

More of a philisophical question: what is the value of readability vs. succinctness in JavaScript

**Comment by erinnachen** [comment]
Things that I learned:
- Something I had not known was the properties that came with event (e.g. .which, .target). I can see it being particularly powerful when multiple actions can occur on a page.
- Event delegation was something that I had no clue what was happening, but had to use when creating new elements on a page. Also the differences between remove, detach, and replaceWith
- .apply is awesome.

Things that I knew:
- I don't know jQuery.

A few questions:
- While I understand the concept of getters and setters is there a reference for what functions are getters vs setters? (probably should just google it come to think of it)
- What is the difference between undefined and null? 
- I do not understand the usage of .then()

**Comment by weilandia** [comment]
What is something you learned that was particularly surprising/interesting?
- Learning about the difference between function declarations and function expressions with regards to hoisting was interesting. What I took away was there are very few benefits to using function declarations and quite a few drawbacks.  One benefit of function declarations might include easier debugging with named functions, but this can be achieved with a hybrid declaration/expression (Named Function Expressions), where you define a variable equal to an anonymous function.

What was something you already knew?
- I already knew a lot of the basics around DOM traversal and manipulation, but it was good to learn some of the nuance in choosing particular methods over others that do similar things.  For example, knowing that .remove() permanently unbinds event handlers, while .detach() preserves them.

Was there anything you feel you still don't understand?
- I'll need to play around with deferreds and promises to fully understand their purpose.

**Comment by brianrip** [comment]
1. What is something you learned that was particularly surprising/interesting?
   I found it useful to understand how to use the the "this" keyword along with chaining methods to           drastically cut the lines of duplicate code I was producing.
2. What was something you already knew?
   Traversing the DOM using html selectors.
3. Was there anything you feel you still don't understand?
   I can still become a little confused with how to properly nest things and select them in the "sibling" structure. Manipulating arrays in JavaScript.

**Comment by hhoopes** [comment]
1. What is something you learned that was particularly surprising/interesting?
   Two different ways to handle events. That there's some flexibility in querying and manipulating, like detaching vs. removing, or addBack
2. What was something you already knew?
   How to find elements in a basic fashion, but not all the relationships like `closest`
3. Was there anything you feel you still don't understand?
   I need to reread event delegation. I've been a little scared of the errors people are getting regarding cross-domain(?) requests, but there seems to be good content regarding that.

**Comment by Tman22** [comment]
1. I didn't know: .is() this would be helpful for testing but also if I need to a truth or false statement for an element on the DOM. Deferreds, though I've read it twice now and think I understand? Event Bubbling, this can be really powerful and actually did help me in ideaBox but I didn't fully understand until I used it.
2. I knew how to manipulating and traversing the DOM. Lovisa lesson in Mod 3 helped me understand how to use parent(), children(), siblings(), hide() and show(). I also knew how to use events and create listeners on the DOM.
3. I don't fully understand Deferreds. I've read it twice and I think I get the general concept but don't see the use case.

**Comment by damwhit** [comment]
1. What is something you learned that was particularly surprising/interesting?
   I learned that you can define a method inside of an object just like any other key/value pair.  So you could do something like var foo = { bar: function() { do stoof} }.
2. What was something you already knew?

I did know that when you assign a function to a variable you no longer have implicit access to it in the global name space like you do when you define the function itself.
1. Was there anything you feel you still don't understand?

Since everything in javascript outside of 5 things is an object I am still not exactly clear on when it is being used functionally vs. object oriented.

**Comment by ksk5280** [comment]
1. What is something you learned that was particularly surprising/interesting?
   - You can easily bind multiple events to a function by just adding the event types in a space-separated string with the .on() method.
   - The event object has a lot of useful properties. It's interesting that you can get the exact coordinates that were clicked.
   - Event delegation seems like a good way to set up event handlers because you can bind to parent elements and target specific element types.
   - form.serialize() takes form input and converts it into query string format which can be passed into the ajax call for data.
2. What was something you already knew?
   - I new most of the basics of JavaScript and jQuery, as well as some of the traversals, how to add, remove, and toggle classes, and a bit about the events, animations and ajax.
3. Was there anything you feel you still don't understand?
   - Namespacing events sounds useful, but I don't understand how to add a namespace to an event.
   - I don't fully understand deferred and all the associated methods. I would need to use it to understand it better and also an example of when would be good to use these methods would help with the understanding as well.

**Comment by brennanholtzclaw** [comment]
1. What is something you learned that was particularly surprising/interesting?
   I was always kind of afraid of `self` in Ruby, but I really liked learning more about `this` and learning how to use it to grab what you need. It made me want to go back and look at some early projects to see where I could use _similar_ ideas to make things smoother.
2. What was something you already knew?
   Manipulating and Traversing the DOM. Andrew gave us a lesson in Mod 2, and we had two more lessons in Mod 3. While I am still not a whiz at it, I feel fairly comfortable figuring out how to get what element I need.
3. Was there anything you feel you still don't understand?
   Getters and Setters seem like a cool concept, but I don't think I really understand how to make the most out of them or exactly when is the best time to use them.

**Comment by scottfirestone** [comment]
1. What is something you learned that was particularly surprising/interesting?
* I'm surprised to have re-read this material and revealed questionable choices from some of my early mod 4 code, mainly using `parents` followed by `children`, instead of `siblings`. Unrelated, I should give my sisters a call. 
* I also failed to serialize form data.

2. What was something you already knew?
* Typically other jQuery traversal / manipulation was understood.

3. Was there anything you feel you still don't understand?
* I don't have a great understanding of Deferreds, and want to use them in a project to better grasp their function.

**Comment by scottfirestone** [comment]
1. What is something you learned that was particularly surprising/interesting?
   I'm surprised to have re-read this material and revealed questionable choices from some of my early mod 4 code, mainly using parents followed by children, instead of siblings. Unrelated, I should give my sisters a call.
   I also failed to serialize form data.
2. What was something you already knew?
   Typically other jQuery traversal / manipulation was understood.
3. Was there anything you feel you still don't understand?
   I don't have a great understanding of Deferreds, and want to use them in a project to better grasp their function.

**Comment by julsfelic** [comment]
1. What is something you learned that was particularly surprising/interesting?
   Maybe this was just an assumption that I made, but jQuery as a whole isn't as big of a library as I thought it was. I also didn't know that they had support for Deferreds (promises) so that is something that will be cool to play around with.
2. What was something you already knew?
   A good majority of it.
3. Was there anything you feel you still don't understand?
   I'm still a little iffy when it comes to fully understanding Deferreds / Promises.

**Comment by adamhundley** [comment]
1. What is something you learned that was particularly surprising/interesting?
- Storing object methods in variables can have unintended consequences for the meaning of 'this'. You can force the meaning of this to be what you want it to be, using the .call() or .apply() method on the function itself. I ran into issues in idea idea box where 'this' had been described at different scopes. This would've been handy!
1. What was something you already knew?
- Variables are a key ingredient in adhering to the Don't Repeat Yourself philosophy. If you need to use a value more than once, you probably want to store it in a variable.
1. Was there anything you feel you still don't understand?
- Nesting. I feel like i constantly tag children and I don't really understand nthchild.
